### ISSUE TITLE
Incorrect CS0411 (type argument inference failure) with three projects when one uses NRTs and one doesn't

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#38898]
**Version Used**: Visual Studio 16.3.1

MSBuild at the command line and VS builds both incorrectly fail with this build output:

> error CS0411: The type arguments for method 'ProjectAStaticClass.GenericMethod\<T>(IBaseInterface\<T>)'
cannot be inferred from the usage. Try specifying the type arguments explicitly.

If you remove `<Nullable>enable</Nullable>` from ProjectA (see below) or add it to ProjectB, ProjectC will stop failing with this bogus error.

The error list sometimes shows this error until you navigate to it. Then it disappears. Other times the VS build fails and the error list stays showing 0 errors. If you specify the type argument, the IDE greys it because it's redundant.

Self-contained repro: [TypeArgumentInferenceFailure.zip](https://github.com/dotnet/roslyn/files/3660231/TypeArgumentInferenceFailure.zip)

Run Repro.ps1. (Adjust the msbuild path if you don't have VS 2019 Enterprise installed at the default location.)

/cc @jcouv, @333fred at @CyrusNajmabadi's recommendation

**Comment by jnm2** [comment]
For convenience:

##### ProjectA.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net472</TargetFramework>
    <LangVersion>latest</LangVersion>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```

##### ProjectA.cs
```cs
public static class ProjectAStaticClass
{
    public static void GenericMethod<T>(IBaseInterface<T> implementation)
    {
    }
}

public interface IBaseInterface<T>
{
}

public interface IInheritingInterfaceWithRecursiveConstraint<T> : IBaseInterface<T>
    where T : GenericClass<T>
{
}

public abstract class GenericClass<T>
{
}

public sealed class DerivedClassSatisfyingRecursiveConstraint : GenericClass<DerivedClassSatisfyingRecursiveConstraint>
{
}
```

##### ProjectB.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net472</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\ProjectA\ProjectA.csproj" />
  </ItemGroup>

</Project>
```

##### ProjectB.cs
```cs
public sealed class ProjectBImplementation : IInheritingInterfaceWithRecursiveConstraint<DerivedClassSatisfyingRecursiveConstraint>
{
}
```

##### ProjectC.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net472</TargetFramework>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\ProjectB\ProjectB.csproj" />
  </ItemGroup>

</Project>
```

##### ProjectC.cs
```cs
public static class ProjectCStaticClass
{
    public static void Foo()
    {
        ProjectAStaticClass.GenericMethod(new ProjectBImplementation());
    }
}
```

**Comment by jcouv** [comment]
Thanks for reporting.
I think I fixed this in 16.4 preview 1 already (PR https://github.com/dotnet/roslyn/pull/38460).
I'll double-check

**Comment by jnm2** [comment]
I should have thought of trying 16.4! It is fixed in preview 4.
